File: day06/java/main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.IntStream;

class Program {
    
    public static void main(String[] args) throws IOException {
        
        var input = Files.readString(Path.of("input"));
        

        System.out.println(solve(4, input));
        System.out.println(solve(14, input));

    }

    public static int solve(int n, String input) {
        return IntStream.range(n, input.length() + 1)
                        .filter(i -> uniqueChars(input.substring(i - n, i)))
                        .findFirst()
                        .orElse(-1);
    }
    
    public static boolean uniqueChars(String str) {
        return str.chars()
                  .distinct()
                  .count() == str.length();
    }
    

}

File: day06/main.py
input = open("input").read()


def solve(input, window_size):
    for i in range(len(input) - window_size):
        window = input[i:i + window_size]
        if (len(window) == len(set(window))):
            return i + window_size
    return None


print(solve(input, 4))
print(solve(input, 14))

input = open("input").read()

def solve(input, window_size):
    return next((i + window_size for i in range(len(input) - window_size) if len(input[i:i + window_size]) == len(set(input[i:i + window_size]))), None)


print(solve(input, 4))
print(solve(input, 14))
